File: MaximalSqaure-DP.java
/*Time Complexity : O(mn)
Space Complexity : 
Did it run on LeetCode : yes
Any Problems: No

Dynamic Programming Solution.
*/
class Solution {
    public int maximalSquare(char[][] matrix) {
     int rows = matrix.length;
        int max=0;   // Create a max variable to compare curr and max. Then return Max*max;
        if(rows==0) return 0;
        int cols = matrix[0].length;
       
        int [] [] dp = new int[rows+1][cols+1]; // Size is taken as rows+1 and cols+1 to take care of edge cases.
        
        
        for(int i=1;i<dp.length;i++)
        {
            for(int j=1;j<dp[0].length;j++)
            {
                if(matrix[i-1][j-1]=='1')
                {
                    dp[i][j]=Math.min(Math.min(dp[i-1][j],dp[i][j-1]),dp[i-1][j-1])+1;
                }
                 max=Math.max(max,dp[i][j]);
            }
           
        }
    return max*max;
    }
}

File: maximal_sqaure_dp.py
"""
Time Complexity : O(m*n) where m is the number of rows and n is the number of cols of the matrix
Space Complexity :  O(m*n) where m is the number of rows and n is the number of cols of the matrix. This space is used for dp matrix

Did this code successfully run on Leetcode : No, Time Limit Exceeded
Any problem you faced while coding this : No
"""


class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if len(matrix) == 0:
            return 0

        m = len(matrix)
        n = len(matrix[0])
        dp = [[0 for _ in range(n+1)] for _ in range(m+1)]
        ans = 0

        for i in range(1, m+1):
            for j in range(1, n+1):
                if matrix[i-1][j-1] == '1':
                    dp[i][j] = 1 + min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1])
                    ans = max(ans, dp[i][j])

        return ans*ans
